src/actions/requests.rs

```diff
@@ -618,37 +636,54 @@ impl<'a> Action<'a> for CodeAction {
 
 impl<'a> RequestAction<'a> for CodeAction {
     type Response = Vec<Command>;
-    fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
+    fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx_: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
         trace!("code_action {:?}", params);
 
-        let ctx = ctx.inited();
+        let ctx = ctx_.inited();
         let file_path = parse_file_path!(&params.text_document.uri, "code_action")?;
 
-        match ctx.previous_build_results.lock().unwrap().get(&file_path) {
-            Some(ref diagnostics) => {
-                let suggestions = diagnostics.iter().filter(|&&(ref d, _)| d.range == params.range).flat_map(|&(_, ref ss)| ss.iter());
-                let mut cmds = vec![];
-                for s in suggestions {
-                    let span = Location {
-                        uri: params.text_document.uri.clone(),
-                        range: s.range,
-                    };
-                    let span = serde_json::to_value(&span).unwrap();
-                    let new_text = serde_json::to_value(&s.new_text).unwrap();
+        let mut cmds = vec![];
+
+        // search for compiler suggestions
+        if let Some(diagnostics) = ctx.previous_build_results.lock().unwrap().get(&file_path) {
+            let suggestions = diagnostics.iter().filter(|&&(ref d, _)| d.range == params.range).flat_map(|&(_, ref ss)| ss.iter());
+            for s in suggestions {
+                let span = Location {
+                    uri: params.text_document.uri.clone(),
+                    range: s.range,
+                };
+                let span = serde_json::to_value(&span).unwrap();
+                let new_text = serde_json::to_value(&s.new_text).unwrap();
+                let cmd = Command {
+                    title: s.label.clone(),
+                    command: "rls.applySuggestion".to_owned(),
+                    arguments: Some(vec![span, new_text]),
+                };
+                cmds.push(cmd);
+            }
+        }
+
+        // search for glob imports
+        // search for a glob in the line
+        let vfs = ctx.vfs.clone();
+        if let Ok(line) = vfs.load_line(&file_path, ls_util::range_to_rls(params.range).row_start) {
+            if line.contains("use ") && line.contains("::*") && !line.contains("prelude::*") {
+                let location = Location::new(params.text_document.uri.clone(), params.range);
+                // call Deglob implementation in order to avoid any false positive of the `contains()`
+                if let Ok(deglob) = Deglob.calculate_workspace_edit(&location, ctx_) {
```

Comment:
Since we're constructing JSON-RPC error response in `Deglob::calculate_workspace_edit`, shouldn't we also do `out.failure_message()` on `Err` here?

Reply:
No, we shouldn't send failures here. The whole point of this call is to see if it succeeds or not. Consider something like

```
/// ```rust
/// use std::mem::*;
/// ```
```

It contains use and ::* so it is a candidate but since it is just a doc comment rls wouldn't be able to replace anything in here. Therefore, I think it should just ignore errors.

Reply:
Ah, yes, my bad. I thought the `Err` is unused, but It's currently used [here](https://github.com/rust-lang-nursery/rls/pull/567/files#diff-3761672bafbe5e9917eb056ea9a482e5R576) instead.